lib/segment/src/types.rs

```diff
@@ -1292,38 +1298,137 @@ pub struct GeoRadius {
 }
 
 impl GeoRadius {
-    pub fn check_point(&self, lon: f64, lat: f64) -> bool {
+    pub fn check_point(&self, point: &GeoPoint) -> bool {
         let query_center = Point::new(self.center.lon, self.center.lat);
-        query_center.haversine_distance(&Point::new(lon, lat)) < self.radius
+        query_center.haversine_distance(&Point::new(point.lon, point.lat)) < self.radius
     }
 }
 
+#[derive(Deserialize)]
+pub struct GeoPolygonShadow {
+    pub rings: Vec<GeoLineString>,
+}
+
 /// Geo filter request
 ///
-/// Matches coordinates inside the polygon, defined by the given coordinates in order.
+/// Matches coordinates inside the polygon, defined by exterior and interiors.
 #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq)]
+#[serde(try_from = "GeoPolygonShadow")]
```

Comment:
why do we need this `GeoPolygonShadow`?

Reply:
I am following this pr: https://github.com/qdrant/qdrant/pull/295. Not sure if there are other better ways but directly implement new() and do validation inside doesn't really work.